aten/src/ATen/cuda/CUDAEvent.h

```diff
@@ -90,20 +119,57 @@ struct AT_CUDA_API CUDAEvent {
 
   // Note: cudaStreamWaitEvent must be called on the same device as the stream.
   // The event has no actual GPU resources associated with it.
-  void block(const CUDAStream& stream) {
+  void block(const CUDAStream& stream) const {
     if (is_created_) {
       CUDAGuard guard(static_cast<int16_t>(stream.device_index()));
       AT_CUDA_CHECK(cudaStreamWaitEvent(stream, event_, 0));
     }
   }
 
+  float elapsed_time(const CUDAEvent& other) const {
+    AT_CHECK(is_created_ && other.isCreated(),
+      "Both events must be recorded before calculating elapsed time.");
+    float time_ms = 0;
+    // raise cudaErrorNotReady if either event is recorded but not yet completed
+    AT_CUDA_CHECK(cudaEventElapsedTime(&time_ms, event_, other.event_));
+    return time_ms;
+  }
+
+  void synchronize() const {
+    if (is_created_) {
+      CUDAGuard guard(static_cast<int16_t>(device_index_));
+      AT_CUDA_CHECK(cudaEventSynchronize(event_));
+    }
+  }
+
+  void ipc_handle(cudaIpcEventHandle_t * handle) {
+    #ifndef __HIP_PLATFORM_HCC__
+      if (!is_created_) {
+        // this CUDAEvent object was initially constructed from flags but event_
+        // is not created yet.
+        createEvent(getCurrentCUDAStream().device_index());
+      }
+      CUDAGuard guard(static_cast<int16_t>(device_index_));
+      AT_CUDA_CHECK(cudaIpcGetEventHandle(handle, event_));
+    #else
+      AT_ERROR("cuIpcGetEventHandle with HIP is not supported");
+    #endif
+  }
+
 private:
   unsigned int flags_ = DEFAULT_FLAGS;
   bool is_created_ = false;
   bool was_recorded_ = false;
   int64_t device_index_ = -1;
```

Comment:
You can make `device_index_` a DeviceIndex instead of `int64_t` and remove all the casts to `int16_t` in things like `CUDAGuard guard(static_cast<int16_t>(device_index_));`